<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8" />
    <meta name="author" content="Brent Pedersen" />
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/selectize.js/0.12.6/css/selectize.bootstrap3.min.css">
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.3.1/css/bootstrap.min.css">
    <script src="https://cdn.plot.ly/plotly-latest.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/selectize.js/0.12.6/js/standalone/selectize.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.3.1/js/bootstrap.bundle.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/protovis/3.3.1/protovis.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/d3/5.16.0/d3.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjs/7.1.0/math.min.js"></script>
    <link rel="stylesheet" type="text/css" href="https://cdn.datatables.net/v/dt/dt-1.10.20/r-2.2.3/sc-2.0.1/sl-1.3.1/datatables.min.css"/>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/datatables/1.10.21/js/jquery.dataTables.min.js"></script>
    <style>

    .big_div {
          height: 80vh;
        }
    
    html,body { height: 100%; margin: 0px; padding: 0px; }
    .form-control.selectize-control {
        padding: 5px 6px 0px;
        height: unset !important;
    }
    .remove-single {
        color: gray !important;
        top: -1px !important;
        font-size: 20px !important;
    }
    </style>
</head>
<body>
    <h1> SEQCOVER </h1>
        <div class="container-fluid p-3">
        <div class="row pt-3 pl-3">
            <div class="form-group", id="regionfilter">
                <label for="regionfilter">Gene/Region Selection</label>
            </div>
        </div>
    <hr>
    <div class="container-fluid p-1">
        <div class="row pt-3">
            <div class="col-6">
                <h4 id="region_header">REGION: </h4>
            </div>
        </div>
        <div class="row">
            <div class="col-12">
                <div class=big_div id="gene_plot"></div>
            </div>
        </div>
        <div class="row ">
            <div class="col-2"></div>
            <div class="col-8">
                <table id="stats_table"></table>
            </div>
            <div class="col-2"></div>
        </div>
        <div class="row pt-5">
            <div class="col-6" id="depth_dist_plot">
            </div>
            <div class="col-6" id="proportion_plot">
            </div>
        </div>
    </div>
</body>

<script src="o.js"></script>

<script>

    var nan = NaN; // hack to support json dumped with NaN values.

    /*
    ------------------------------------------------------------------------------------------------------------------
                                                        SETUP 
    ------------------------------------------------------------------------------------------------------------------
    */

    //Get gene/region names and add them to the available regions form 
    var region_index_map = {};
    var select = $('<select class="form-control" id="regionselect"></select>');
    var index_map = {};
    for (index in plot_data) {

        //Get region gene info
        var symbol = plot_data[index].symbol;
        var chrom = plot_data[index].unioned_transcript.chr;
        var genomic_positions = plot_data[index].plot_coords.g;
        var start = Math.min.apply(null,genomic_positions.filter(Boolean));
        var end = Math.max(...genomic_positions);

        //Get region string
        var region_string = symbol + " "+ chrom + ":" + start + "-" + end

        //Store the index of the region in the plot data object.
        region_index_map[region_string] = index 
        
        //Add region to select options
        var option = $("<option>" + region_string + "</option>");
        select.append(option)
    }

    //Add new select options to regionfilter div
    $("#regionfilter").append(select);


    /*
    ------------------------------------------------------------------------------------------------------------------
                                                      Plot Functions 
    ------------------------------------------------------------------------------------------------------------------
    */

    //by base depth plot layout
    function get_gene_plot_layout(gene) {

        var mid = Math.round(gene.plot_coords.x.length / 2)

        var layout = {
            grid: {
                rows: 3,
                columns: 1,
            },
            autosize: true,
            margin: {t: 30, pad: 0, l: 150},
            xaxis: {
                tickmode: "array",
                tickvals: [gene.plot_coords.x[0], gene.plot_coords.x[mid], gene.plot_coords.x[gene.plot_coords.x.length - 10]],
                ticktext: [gene.plot_coords.g[0], gene.plot_coords.g[mid], gene.plot_coords.g[gene.plot_coords.x.length - 10]],
                title: "Chromosome " + String(gene.unioned_transcript.chr).replace("chr","")
            },
            yaxis: {
                title: "Depth",
                domain: [0.55,0.90]
            },
            yaxis2: {
                title: "Merged<br>Transcripts",
                range: [-8,8],
                showlegend: false,
                tickvals: [1],
                showticklabels: false,
                ticktext: gene.unioned_transcript.transcript,
                domain: [0.4,0.55],

            },
            yaxis3: {
                range: [0,2],
                showlegend: false,
                domain: [0.0,0.40]
            },
            hovermode: 'closest',
            showlegend: true,
            legend: {
                xanchor: "right",
                yanchor: "top",
                y: 1,
                x: 1,
                orientation: "h",
                borderwidth: 1,
                bordercolor: '#eeeeee'
            },
        };

        return(layout)

    };


    //Get a by position trace per sample of depth 
    function get_by_position_depth_trace(gene) {

        var traces = [];

        for (sample in gene.plot_coords.depths) {
            var dp = gene.plot_coords.depths[sample]
            dp = dp.map(function(v) { return v < -1000 ? NaN : v})

            var trace = {x: gene.plot_coords.x, text: gene.plot_coords.g, y: dp, 
                type: 'scatter', mode:'lines', name: sample, line: {width: 1},
                hovertemplate: '<b>position</b>:%{text}<br><b>depth</b>:%{y}<br>(debug) x: %{x}',
                hoverinfo:"text",
                yaxis: "y",
            };

            traces.push(trace);
        };

        return(traces)

    };


    //Get the unioned transcript shapes
    function get_unioned_transcript_shapes(gene) {

        let transcript_color = 'rgb(100, 100, 100)';
        let cds_color = 'rgb(100, 100, 255)';
        let exon_color = 'rgb(200, 100, 255)';
        var yoff = 1;
        var shapes = [];
        var tr = gene.unioned_transcript;

        shapes.push({type: 'rect', x0: tr.txstart, x1: tr.txstop, y0: yoff-0.1, y1: yoff+0.1, line: {color: transcript_color}, yref: "y2" })
        shapes.push({type: 'rect', x0: tr.cdsstart, x1: tr.cdsstop, y0: yoff-1, y1: yoff+1, line: {color: cds_color},  yref: "y2"})

        tr.position.forEach(p => shapes.push({type: "rect", x0: p[0], x1: p[1], y0: yoff-2, y1:yoff+2 ,
                                              fillcolor: exon_color, line: {color:exon_color}, yref: "y2"}))

        return(shapes)

    };


    //Get a list of transcript information to use as a custoomdata imput for plotly hover data
    function get_custom_data_list(transcript, exonStart, exonEnd, exonNumber, gposition_list ) {

        custdata = []

        //Format = transcript id, strand, txstart, txend, cdsstart, cdsend, exon/utr start, exon/utr end, exon number or utr type 
        custdata.push( transcript.transcript ) // Transcript id
        custdata.push( transcript.strand > 0 ? "+" : "-") // Strand 
        custdata.push( gposition_list[transcript.txstart] ) // transcript start 
        custdata.push( gposition_list[transcript.txend] ) // transcript end
        custdata.push( gposition_list[transcript.cdsstart] ) // cds start 
        custdata.push( gposition_list[transcript.cdsend] ) // cds end
        custdata.push( gposition_list[exonStart] ) // Start of UTR
        custdata.push( gposition_list[exonEnd] ) //END of UTR
        custdata.push( exonNumber ) // Exon Number or UTR type 

        return(custdata)

    }; 
    
    //Get the transcript shapes for the current region
    function get_transcript_shapes(gene) {
        
        let symbol = gene.symbol
        let top = 10;
        let bottom = 8;
        let exonOffset = 0.25;
        let utrOffset = 0.1;
        let shapes = [];
        let tick_labels = [];
        let tick_vals = [];
        let strandArray = [];
        let x_values = []
        let y_values = []
        let customData = []
        let text_symbol = []

        for (transcript of gene.transcripts) {

            // Transcript 
            shapes.push({type: "line", x0: transcript.txstart, x1: transcript.txend, 
                         y0: bottom + 1, y1: bottom + 1, line_color: "black",yref: "y3"}) 

            //Strand info
            let strand_interval = Math.round((transcript.txend - transcript.txstart) / 20)
            let strand_marker = transcript.strand > 0 ? ">" : "<";
            for (i = transcript.cdsstart; i < transcript.cdsend; i++) {
                if (i % strand_interval == 0) {

                    strandArray.push({x: i, y: bottom + 1.0, text: strand_marker, font: {color: "black", size: 16}, opacity: 0.7, showarrow: false, yref: "y3"})

                };
            };


            //Get Start UTRs 
            let utr1 = {type: "rect", x0:transcript.txstart, x1:transcript.cdsstart, 
                        y0: bottom + (1 - utrOffset), y1: bottom + (1 + utrOffset),
                        line_color: "black", fillcolor: "cyan", yref: "y3"}

            //Add UTR info to customData
            customData.push(get_custom_data_list(transcript, transcript.txstart, transcript.cdsstart, transcript.strand > 0 ? "5'-UTR" : "3'-UTR", gene.plot_coords.g)) 
            text_symbol.push(symbol) 

            //Add UTRs to shapes
            shapes.push(utr1)

            // Add UTR plot coordinates for hover info 
            x_values.push((transcript.txstart + ((transcript.cdsstart - transcript.txstart) / 2)))
            y_values.push(bottom + 1)


            //Get End UTR 
            let utr2 = {type: "rect", x0:transcript.cdsend, x1:transcript.txend, 
                        y0: bottom + (1 - utrOffset), y1: bottom + (1 + utrOffset),
                        line_color: "cyan", fillcolor: "cyan", yref: "y3"}

            //Add UTR info to customData
            customData.push(get_custom_data_list(transcript, transcript.cdsend, transcript.txend, transcript.strand > 0 ? "3'-UTR" : "5'-UTR", gene.plot_coords.g)) 
            text_symbol.push(symbol) 

            //Add UTRs to shapes
            shapes.push(utr2)

            // Add UTR plot coordinates for hover info 
            x_values.push((transcript.cdsend + ((transcript.txend - transcript.cdsend) / 2)))
            y_values.push(bottom + 1)

            // Get exon numbers
            var exonCount = transcript.position.length
            var exonNumbers = transcript.strand > 0 ? pv.range(1,exonCount + 1) : pv.range(exonCount, 0, -1) 

            //Add exon shapes
            exon_2d_array = transcript.position.sort(function(a,b) {return a[0] - b[0]})
            for (index in exon_2d_array) {
                
                let exonPos = transcript.position[index]

                //If the exon start position is past the end of the cds region, skip this exon (In UTR region)
                if (exonPos[0] > transcript.cdsend) {
                    exonCount -= 1
                    exonNumbers = transcript.strand > 0 ? pv.range(1,exonCount + 1) : pv.range(exonCount, 0, -1) 
                    continue
                }

                //If the exon end position is before the start of the cds region, skip this exon (In UTR region)
                if (exonPos[1] < transcript.cdstart) {
                    exonCount -= 1
                    exonNumbers = transcript.strand > 0 ? pv.range(1,exonCount + 1) : pv.range(exonCount, 0, -1) 
                    continue
                }


                //Get exon start and end positions
                let exonstart = exonPos[0] > transcript.cdsstart ? exonPos[0] : transcript.cdsstart
                let exonend = exonPos[1] < transcript.cdsend ? exonPos[1] : transcript.cdsend 

                //Add exon to shapes
                shapes.push({type: "rect", x0: exonstart, x1: exonend,  y0: bottom + (1 - exonOffset), 
                             y1: bottom + (1 + exonOffset), line_color: "black", fillcolor: "black", yref: "y3"}) 

                //Add exon info to customData
                customData.push(get_custom_data_list(transcript, exonstart, exonend, exonNumbers[index], gene.plot_coords.g)) 
                text_symbol.push(symbol) 

                // Add exon plot coordinates for hover info
                x_values.push((exonstart + ((exonend - exonstart) / 2)))
                y_values.push(bottom + 1)

            };
            
                
            //Get tick labels
            tick_labels.push(transcript.transcript)
            tick_vals.push(bottom + 1)

            //update bottom 
            bottom = bottom - 1 
        };

        //Format = transcript id, strand, txstart, txend, cdsstart, cdsend, exon/utr start, exon/utr end, exon number or utr type 
        let trace = {x: x_values, y: y_values, text: text_symbol, 
                type: 'scatter', mode:'markers', customdata: customData, name: "transcripts",
                opacity: 0.0,showlegend: false, yaxis: "y3", marker: { color: "darkcyan", size: 20}, 
                hovertemplate: '<b>Gene</b>: %{text}<br><b>Transcript ID</b>: %{customdata[0]}<br><b>Strand</b>: %{customdata[1]}<br><b>TX Start</b>: %{customdata[2]}<br><b>TX End</b>: %{customdata[3]}<br><b>CDS Start</b>: %{customdata[4]}<br><b>CDS End</b>: %{customdata[5]}<br><b>Exon Start</b>: %{customdata[6]}<br><b>Exon End</b>: %{customdata[7]}<br><b>Exon Number</b>: %{customdata[8]}<br>',
            };

        return({"shapes": shapes, "top": top, "bottom": bottom, "tick_labels": tick_labels, "tick_vals": tick_vals, "strand_annotation": strandArray, "trace": trace})

    };


    //Get descriptive stats info for stats table
    function get_stats_table(gene, low_depth = 10) {

        var dataRows = []
        var columnNames = [{"title": "Sample", width: "20%"},
                           {"title": "Mean Depth", width: "10%"},
                           {"title": "Median Depth", width: "10%"},
                           {"title": "STD", width: "10%"},
                           {"title": "Min Depth", width: "10%"},
                           {"title": "Max Depth", width: "10%"},
                           {"title": "Low Coverage Sites<br>( <=" + low_depth + " )", width: "30%"}]
    
        for (sample in gene.plot_coords.depths) {

            // Depths 
            let dp = gene.plot_coords.depths[sample]

            //Convert null values to NaN 
            dp = dp.map(function(v) { return v < -1000 ? NaN : v})
            
            //Remove NaNs
            dp = dp.filter(value => !Number.isNaN(value));

            //Get depth descriptive statistics
            let dp_min = Math.min(...dp)
            let dp_max = Math.max(...dp)
            let dp_mean = math.round(math.mean(dp),3)
            let dp_median = math.median(dp)
            let dp_std = math.round(math.std(dp),3)
            let dp_low = math.sum(dp.map(function(v) { return v <= low_depth ? 1: 0})) 

            dataRows.push([sample, dp_mean, dp_median, dp_std, dp_min, dp_max, dp_low])
        };
        return({rowData: dataRows, colNames: columnNames})
    };


    //Get the traces for each sample depth distribution
    function get_depth_distribution_traces(gene) {

        let traces = []

        for (sample in gene.plot_coords.depths) {

            // Depths 
            let dp = gene.plot_coords.depths[sample]

            //Convert null values to NaN 
            dp = dp.map(function(v) { return v < -1000 ? NaN : v})
            
            //Remove NaNs
            dp = dp.filter(value => !Number.isNaN(value));

            let depth_map = {}
            let depth_values = new Set()
            let dp_count = 0 
            //get the number of 
            for (depth of dp) {
                
                dp_count += 1
                
                depth_values.add(parseInt(depth))

                if (!(depth in depth_map)) {
                    depth_map[depth] = 1

                } else {
                    depth_map[depth] += 1
                };
            };

            let x_values = []
            let y_values = []
            //Calculate proportion scores
            for (key of [...depth_values].sort(function (a,b) {return (a - b)})) {

                //Add x and y valuess
                x_values.push(key)
                y_values.push(depth_map[key])

            };

            let trace = {x: x_values, y: y_values,  mode: "lines", 
                         name: sample, line: {width: 1}, 
                         hovertemplate: "<b>Depth</b>:%{x}<br><b>Count</b>:%{y}"}

            traces.push(trace)
        };
        
        return(traces)

    };


    //Get the depth distribution layout
    function get_depth_distribution_layout() {

        var layout = {
            autosize: true,
            title: "Depth Distribution per Sample",
            xaxis: {
                title: "Depth"
            },
            yaxis: {
                title: "Count",
                domain: [0.0,0.9],
            },
            hovermode: 'closest',
            showlegend: true,
            legend: {
                xanchor: "right",
                yanchor: "top",
                y: 1,
                x: 1,
                orientation: "h",
                borderwidth: 1,
                bordercolor: '#eeeeee'
            },
        };
        return(layout)
    };


    //per Sample depth proportion traces
    function get_proportion_trace(gene) {

        let traces = []

        for (sample in gene.plot_coords.depths) {

            // Depths 
            let dp = gene.plot_coords.depths[sample]

            //Convert null values to NaN 
            dp = dp.map(function(v) { return v < -1000 ? NaN : v})
            
            //Remove NaNs
            dp = dp.filter(value => !Number.isNaN(value));

            let depth_map = {}
            let depth_values = new Set()
            let dp_count = 0 
            //get the number of 
            for (depth of dp) {
                
                dp_count += 1
                
                depth_values.add(depth)

                if (!(depth in depth_map)) {
                    depth_map[depth] = 1

                } else {
                    depth_map[depth] += 1
                };
            };

            let x_values = []
            let y_values = []
            let counter = 0
            //Calculate proportion scores
            for (key of [...depth_values].sort(function(a,b) {return b-a})) {

                //increment counter
                counter += depth_map[key]

                //Add x and y valuess
                x_values.push(key)
                y_values.push(counter / dp_count)

            };

            let trace = {x: x_values, y: y_values,  mode: "lines", 
                         name: sample, line: {width: 1}, 
                         hovertemplate: "<b>Depth</b>:%{x}<br><b>Proportion</b>:%{y}"}

            traces.push(trace)
        };
        
        return(traces)
    };


    //by proprotion plot layout
    function get_proportion_layout() {

        var layout = {
            autosize: true,
            title: "Depth Proportion per Sample",
            xaxis: {
                title: "Depth"
            },
            yaxis: {
                title: "Proportion",
                domain: [0.0,0.9],
            },
            hovermode: 'closest',
            showlegend: true,
            legend: {
                xanchor: "right",
                yanchor: "top",
                y: 1,
                x: 1,
                orientation: "h",
                borderwidth: 1,
                bordercolor: '#eeeeee'
            },
        };
        return(layout)
    };
    

    /*
    ------------------------------------------------------------------------------------------------------------------
                                                     Plot Controllers 
    ------------------------------------------------------------------------------------------------------------------
    */

    //Per base depth plot with transcript annotations
    function plot_per_base_depth(gene) {

        //Get gene specific layout
        plot_gene_layout = get_gene_plot_layout(gene)

        //Get gene specific depth traces
        by_position_depth_traces = get_by_position_depth_trace(gene)

        //Add empty subplots 
        by_position_depth_traces.push({yaxis: "y2"})


        // get unionized transcript shapes
        unioned_transcript_shapes = get_unioned_transcript_shapes(gene)
        
        // Get per transcript annotation shapes
        transcript_map = get_transcript_shapes(gene)
        by_position_depth_traces.push(transcript_map.trace)

        //Add shapes to layout 
        plot_gene_layout.shapes = unioned_transcript_shapes.concat(transcript_map.shapes)

        //Adjuste subplot 3 yaxis ranges based on the number of annotated  transcripts 
        plot_gene_layout.yaxis3.range = [transcript_map.bottom, transcript_map.top]

        //Add subplot 3 yaxis tick labels (transcript ids)
        plot_gene_layout.yaxis3.tickvals = transcript_map.tick_vals
        plot_gene_layout.yaxis3.ticktext = transcript_map.tick_labels

        //Add subplot 3 strand annotations 
        plot_gene_layout.annotations = transcript_map.strand_annotation

        //Plot by base depth 
        Plotly.newPlot("gene_plot", by_position_depth_traces, plot_gene_layout)

    };


    //Add a jQuery data table with descriptive stats
    function plot_stats_table(gene) {

        //Remove previous data table if it exists
        if ($.fn.DataTable.isDataTable('#stats_table')) {
            let table = $('#stats_table').DataTable()
            table.destroy();
        };

        //Get row data and column names
        rows_and_columns = get_stats_table(gene)

        //Create data table
        jQuery("#stats_table").DataTable({
            destory: true,
            data: rows_and_columns["rowData"],
            columns: rows_and_columns["colNames"],
            paging: true,
            scroller: true,
            scrollY: '60vh',
            scrollX: true,
            scrollCollapse: true,
            select: {style: "os"},
        });

    };


    //Depth Distribution plot using per depth count 
    function plot_depth_distribution(gene) {

        //Get depth distribution plot trace per sample 
        var depth_traces = get_depth_distribution_traces(gene)

        //Get the plot layout for the depth plot
        var depth_dist_layout = get_depth_distribution_layout()

        Plotly.newPlot("depth_dist_plot", depth_traces, depth_dist_layout)

    };



    /*
    ------------------------------------------------------------------------------------------------------------------
                                                     Event Handling 
    ------------------------------------------------------------------------------------------------------------------
    */

    //Controller function for generating gene/region sepecific plots
    function generate_plots(selected_region) {

        //Update region specific header  
        var header = document.getElementById("region_header")
        header.innerHTML = selected_region
        
        //Index of plot in plot_data and gene data
        var plotIndex = region_index_map[selected_region];
        var selectedGene = plot_data[plotIndex];

        //Plot per base depths 
        plot_per_base_depth(selectedGene)

        //Add a descriptive statistics table
        plot_stats_table(selectedGene)

        //Plot depth distribution distrubtion 
        plot_depth_distribution(selectedGene)
        
    };


    //Load first region
    jQuery(document).ready(function(){

        //Plot first gene
        for (region in region_index_map){
            if (region_index_map[region] == 0) {
                generate_plots(region)
                break
            };
        };


    });


    //Select gene/region change
    $("#regionselect").on('change', function(e) {
        var selectedValue = this.value

        //Plot the selected region
        generate_plots(selectedValue)

    });


    

</script>

</body>
</html>
